Compiler panic on building Redox

I tried to build current Redox OS HEAD with last available rustc-nightly and got an ICE.

## Meta
`rustc --version --verbose`:
```
rustc 1.9.0-nightly (241a9d0dd 2016-04-05)
binary: rustic
commit-hash: 241a9d0ddf99fd40d273c615e9b1e8ce6052d94a
commit-date: 2016-04-05
host: x86_64-apple-darwin
release: 1.9.0-nightly
```

Backtrace:
```
❯❯❯ make qemu                                                                                                                                                ⏎ master
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/libstd.rlib libstd/src/lib.rs -L native=libc/lib/
error: internal compiler error: ../src/librustc_metadata/encoder.rs:220: encode_symbol: id not found 49044
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', ../src/libsyntax/errors/mod.rs:575
stack backtrace:
   1:        0x110b66978 - std::sys::backtrace::tracing::imp::write::h714760a4c8c0cdd8
   2:        0x110b72dc5 - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hff309ab1d83ffd90
   3:        0x110b728fd - std::panicking::default_hook::h08ad3bb09872855b
   4:        0x110b36196 - std::sys_common::unwind::begin_unwind_inner::h406d5f1a330b854b
   5:        0x110113cea - std::sys_common::unwind::begin_unwind::hb10e2b9ef92fd4a6
   6:        0x110113b19 - syntax::errors::Handler::bug::h112319de7e910c9e
   7:        0x10f494c9c - rustc::session::opt_span_bug_fmt::_$u7b$$u7b$closure$u7d$$u7d$::h2c926d3e6a2eb430
   8:        0x10f494b01 - rustc::session::opt_span_bug_fmt::he2c7fbeab8bf3d1d
   9:        0x10f4a36a9 - rustc::session::bug_fmt::h9f67e9cd7aa5a073
  10:        0x10cab64f0 - rustc_metadata::encoder::encode_symbol::h165ac8a162612975
  11:        0x10cac310b - rustc_metadata::encoder::encode_info_for_item::hfab1a46f9bf30c3c
  12:        0x10cac9c71 - rustc::front::map::Map::with_path_next::h6ba163a93b4eab46
  13:        0x10cac9ca4 - rustc::front::map::Map::with_path_next::h6ba163a93b4eab46
  14:        0x10cac9ca4 - rustc::front::map::Map::with_path_next::h6ba163a93b4eab46
  15:        0x10caccd4e - _<encoder..EncodeVisitor<'a, 'b, 'c, 'tcx> as rustc_front..intravisit..Visitor<'tcx>>::visit_item::h1cde6db68ffd78b8
  16:        0x10cad3f5d - rustc_metadata::encoder::encode_metadata_inner::hcfb24759120a6c88
  17:        0x10cacfb19 - rustc_metadata::encoder::encode_metadata::ha3e422357afb6bf6
  18:        0x10cb2c07f - rustc_metadata::csearch::_<impl rustc..middle..cstore..CrateStore<'tcx> for cstore..CStore>::encode_metadata::h88a2f338ac9ea48a
  19:        0x10ccafee2 - rustc_trans::base::trans_crate::_$u7b$$u7b$closure$u7d$$u7d$::h6042c55692ac48ec
  20:        0x10cca1dfd - rustc_trans::base::trans_crate::h45d3e3b1384148fb
  21:        0x10c24fce0 - rustc_driver::driver::phase_4_translate_to_llvm::h8d5ab3a4d94eae6a
  22:        0x10c24e434 - rustc_driver::driver::compile_input::_$u7b$$u7b$closure$u7d$$u7d$::hd32acfaba89541b2
  23:        0x10c24b02d - rustc_driver::driver::phase_3_run_analysis_passes::_$u7b$$u7b$closure$u7d$$u7d$::h1692b932d4d7b15b
  24:        0x10c2451a5 - rustc::ty::context::TyCtxt::create_and_enter::h926ea4fe45f15d00
  25:        0x10c241ba5 - rustc_driver::driver::phase_3_run_analysis_passes::h3c035e5cf4758407
  26:        0x10c215af1 - rustc_driver::driver::compile_input::h650fe5b01cb8d74d
  27:        0x10c1fc04f - rustc_driver::run_compiler::h68d23e0e9b7b247d
  28:        0x10c1f93e2 - std::sys_common::unwind::try::try_fn::h67fde221a73148bc
  29:        0x110b63fdb - __rust_try
  30:        0x110b63f63 - std::sys_common::unwind::inner_try::h4e97625a08807651
  31:        0x10c1f9c79 - _<F as std..boxed..FnBox<A>>::call_box::hc8936fa120642c49
  32:        0x110b71c88 - std::sys::thread::Thread::new::thread_start::h74af400293164137
  33:     0x7fff9f86399c - _pthread_body
  34:     0x7fff9f863919 - _pthread_start

make: *** [build/i386-unknown-redox/debug/libstd.rlib] Error 101
```

Same here.  `rustc -V` responds with `rustc 1.9.0-nightly (7979dd608 2016-04-07)` and `make all` produces:

```
mkdir -p build/i386-unknown-redox/debug
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/libcore.rlib rust/src/libcore/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/libio.rlib crates/io/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/liballoc_system.rlib liballoc_system/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/liballoc.rlib rust/src/liballoc/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/librustc_unicode.rlib rust/src/librustc_unicode/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/libcollections.rlib rust/src/libcollections/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/librand.rlib rust/src/librand/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/libsystem.rlib crates/system/lib.rs
RUST_BACKTRACE=1 rustc --target=i386-unknown-redox.json -L build/i386-unknown-redox/debug -C no-prepopulate-passes -C no-stack-check -C opt-level=2 -Z no-landing-pads -A dead_code -C ar=i386-elf-ar -C linker=i386-elf-gcc -o build/i386-unknown-redox/debug/libstd.rlib libstd/src/lib.rs -L native=libc/lib/
error: internal compiler error: ../src/librustc_metadata/encoder.rs:218: encode_symbol: id not found 49044
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', ../src/libsyntax/errors/mod.rs:575
stack backtrace:
   1:        0x10b8e4838 - std::sys::backtrace::tracing::imp::write::h714760a4c8c0cdd8
   2:        0x10b8f0c75 - std::panicking::default_hook::_$u7b$$u7b$closure$u7d$$u7d$::hff309ab1d83ffd90
   3:        0x10b8f07af - std::panicking::default_hook::h08ad3bb09872855b
   4:        0x10b8b5396 - std::sys_common::unwind::begin_unwind_inner::h406d5f1a330b854b
   5:        0x10aeadc3a - std::sys_common::unwind::begin_unwind::hb10e2b9ef92fd4a6
   6:        0x10aeada69 - syntax::errors::Handler::bug::h112319de7e910c9e
   7:        0x10a26babc - rustc::session::opt_span_bug_fmt::_$u7b$$u7b$closure$u7d$$u7d$::h2c926d3e6a2eb430
   8:        0x10a26b921 - rustc::session::opt_span_bug_fmt::he2c7fbeab8bf3d1d
   9:        0x10a282119 - rustc::session::bug_fmt::h9f67e9cd7aa5a073
  10:        0x10754e9c0 - rustc_metadata::encoder::encode_symbol::h165ac8a162612975
  11:        0x10755a1d3 - rustc_metadata::encoder::encode_info_for_item::h60f93cd3dd34a5cf
  12:        0x10756a42d - rustc_metadata::encoder::encode_metadata_inner::ha8580aa6ef04664a
  13:        0x107565fe9 - rustc_metadata::encoder::encode_metadata::h4120aac9f1041ba6
  14:        0x1075c1a1f - rustc_metadata::csearch::_<impl rustc..middle..cstore..CrateStore<'tcx> for cstore..CStore>::encode_metadata::he92619b644b6d919
  15:        0x107947262 - rustc_trans::base::trans_crate::_$u7b$$u7b$closure$u7d$$u7d$::h6042c55692ac48ec
  16:        0x107939f0d - rustc_trans::base::trans_crate::h45d3e3b1384148fb
  17:        0x106fcf3aa - rustc_driver::driver::phase_4_translate_to_llvm::h8d5ab3a4d94eae6a
  18:        0x106fcdb5c - rustc_driver::driver::compile_input::_$u7b$$u7b$closure$u7d$$u7d$::hd32acfaba89541b2
  19:        0x106fca4ed - rustc_driver::driver::phase_3_run_analysis_passes::_$u7b$$u7b$closure$u7d$$u7d$::h8148bf3bd7eb17b2
  20:        0x106fc40a5 - rustc::ty::context::TyCtxt::create_and_enter::h963727672a0d2ef1
  21:        0x106fc0aac - rustc_driver::driver::phase_3_run_analysis_passes::h035bf2dff707c080
  22:        0x106f946e1 - rustc_driver::driver::compile_input::h650fe5b01cb8d74d
  23:        0x106f7aeff - rustc_driver::run_compiler::h68d23e0e9b7b247d
  24:        0x106f78292 - std::sys_common::unwind::try::try_fn::h67fde221a73148bc
  25:        0x10b8e1fcb - __rust_try
  26:        0x10b8e1f53 - std::sys_common::unwind::inner_try::h4e97625a08807651
  27:        0x106f78b29 - _<F as std..boxed..FnBox<A>>::call_box::hc8936fa120642c49
  28:        0x10b8efb48 - std::sys::thread::Thread::new::thread_start::h74af400293164137
  29:     0x7fff8626f99c - _pthread_body
  30:     0x7fff8626f919 - _pthread_start

Makefile:434: recipe for target 'build/i386-unknown-redox/debug/libstd.rlib' failed
make: *** [build/i386-unknown-redox/debug/libstd.rlib] Error 101
```

--- a/src/librustc_trans/callee.rs
+++ b/src/librustc_trans/callee.rs
@@ -582,14 +582,18 @@ fn get_fn<'a, 'tcx>(ccx: &CrateContext<'a, 'tcx>,
         debug!("get_fn: not casting pointer!");
 
         attributes::from_fn_attrs(ccx, attrs, llfn);
-        if let Some(id) = local_item {
+        if local_item.is_some() {
             // FIXME(eddyb) Doubt all extern fn should allow unwinding.
             attributes::unwind(llfn, true);
-            ccx.item_symbols().borrow_mut().insert(id, sym);
         }
 
         llfn
     };
+
+    // Always insert into item_symbols, in case this item is exported.
+    if let Some(id) = local_item {
+        ccx.item_symbols().borrow_mut().insert(id, sym);
+    }
 
     ccx.instances().borrow_mut().insert(instance, llfn);
 

--- a/src/test/auxiliary/foreign_lib.rs
+++ b/src/test/auxiliary/foreign_lib.rs
@@ -9,6 +9,7 @@
 // except according to those terms.
 
 #![crate_name="foreign_lib"]
+
 #![feature(libc)]
 
 pub mod rustrt {
@@ -17,5 +18,31 @@ pub mod rustrt {
     #[link(name = "rust_test_helpers")]
     extern {
         pub fn rust_get_test_int() -> libc::intptr_t;
+    }
+}
+
+pub mod rustrt2 {
+    extern crate libc;
+
+    extern {
+        pub fn rust_get_test_int() -> libc::intptr_t;
+    }
+}
+
+pub mod rustrt3 {
+    // Different type, but same ABI (on all supported platforms).
+    // Ensures that we don't ICE or trigger LLVM asserts when
+    // importing the same symbol under different types.
+    // See https://github.com/rust-lang/rust/issues/32740.
+    extern {
+        pub fn rust_get_test_int() -> *const u8;
+    }
+}
+
+pub fn local_uses() {
+    unsafe {
+        let x = rustrt::rust_get_test_int();
+        assert_eq!(x, rustrt2::rust_get_test_int());
+        assert_eq!(x as *const _, rustrt3::rust_get_test_int());
     }
 }

--- a/src/test/run-pass/foreign-dupe.rs
+++ b/src/test/run-pass/foreign-dupe.rs
@@ -8,41 +8,18 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-// calling pin_thread and that's having weird side-effects.
+// aux-build:foreign_lib.rs
 
-#![feature(libc)]
+// Check that we can still call duplicated extern (imported) functions
+// which were declared in another crate. See issues #32740 and #32783.
 
-mod rustrt1 {
-    extern crate libc;
 
-    #[link(name = "rust_test_helpers")]
-    extern {
-        pub fn rust_get_test_int() -> libc::intptr_t;
-    }
-}
-
-mod rustrt2 {
-    extern crate libc;
-
-    extern {
-        pub fn rust_get_test_int() -> libc::intptr_t;
-    }
-}
-
-mod rustrt3 {
-    // Different type, but same ABI (on all supported platforms).
-    // Ensures that we don't ICE or trigger LLVM asserts when
-    // importing the same symbol under different types.
-    // See https://github.com/rust-lang/rust/issues/32740.
-    extern {
-        pub fn rust_get_test_int() -> *const u8;
-    }
-}
+extern crate foreign_lib;
 
 pub fn main() {
     unsafe {
-        let x = rustrt1::rust_get_test_int();
-        assert_eq!(x, rustrt2::rust_get_test_int());
-        assert_eq!(x as *const _, rustrt3::rust_get_test_int());
+        let x = foreign_lib::rustrt::rust_get_test_int();
+        assert_eq!(x, foreign_lib::rustrt2::rust_get_test_int());
+        assert_eq!(x as *const _, foreign_lib::rustrt3::rust_get_test_int());
     }
 }

